apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,378 @@
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// +genclient
+// +genclient:namespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope="Namespaced",categories={fleet,fleet-placement}
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// StagedUpdateStrategy defines a reusable strategy that specifies the stages and the sequence
+// in which the application will be updated on the member clusters.
+type StagedUpdateStrategy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Stage configurations for each update stage.
+	// +required
+	Stages []StageConfig `json:"stages"`
+
+	// TODO: Add alerting configuration.
+}
+
+// StagedUpdateStrategyList contains a list of StagedUpdateStrategy.
+// +kubebuilder:resource:scope="Namespaced"
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type StagedUpdateStrategyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []StagedUpdateStrategy `json:"items"`
+}
+
+// StageConfig describes a single update stage group configuration.
+// The clusters in each stage are updated sequentially for now.
+// We will stop the update if any of the updates fail.
+type StageConfig struct {
+	// The name of the stage. This MUST be unique within the same StagedUpdateStrategy.
+	// +required
+	Name string `json:"name"`
+
+	// LabelSelector is a label query over all the joined member clusters. Clusters matching the query are selected.
+	// We don't check if there are overlap between clusters selected by different stages.
+	LabelSelector *metav1.LabelSelector `json:"labelSelector"`
+
+	// The label key used to sort the selected clusters.
+	// The clusters within the stage are updated sequentially following the rule below:
+	//   - primary: Ascending order based on the value of the label key, interpreted as integers if present.
+	//   - secondary: Ascending order based on the name of the cluster if the label key is absent.
+	// +optional
+	SortingLabelKey *string `json:"sortingLabelKey,omitempty"`
+
+	// The collection of tasks that each stage needs to complete successfully before moving to the next stage.
+	// Each task is executed in parallel and they can not be more than one task of the same type.
+	// +optional
+	AfterStageTasks []AfterStageTask `json:"afterStageTasks,omitempty"`
+}
+
+// AfterStageTask is the collection of post stage tasks that ALL need to be completed before we can move to the next stage.
+type AfterStageTask struct {
+	// The type of the after stage task.
+	// +kubebuilder:validation:Enum=TimedWait;Approval
+	// +required
+	Type AfterStageTaskType `json:"type"`
+
+	// The prefix of the name of the StageApprovalRequest we need to create to get the approval for the next stage.
+	// We create the full name of the approval request by concatenating the prefix to the stagedUpdateRunName and the stageName.
```

Comment:
Could you provide the approval request name template and example?
<prefix>-<runName>-<stageName>, e.g., "myPrefix-run20240908-stage1"

BTW, do we really need the prefix? Isn't "run20240908-stage1" good enough? When is the prefix useful?

Reply:
Unresovled?

Reply:
removed this field